diff --git a/src/libsodium/crypto_generichash/blake2/ref/blake2b-ref.c b/src/libsodium/crypto_generichash/blake2/ref/blake2b-ref.c
index bffac86..c3c35ac 100644
--- a/src/libsodium/crypto_generichash/blake2/ref/blake2b-ref.c
+++ b/src/libsodium/crypto_generichash/blake2/ref/blake2b-ref.c
@@ -29,6 +29,9 @@ typedef unsigned uint128_t __attribute__ ((mode(TI)));
 # endif
 #endif
 
+extern void bsod(const char*,int);
+#define abort() bsod(__FILE__,__LINE__)
+
 static blake2b_compress_fn blake2b_compress = blake2b_compress_ref;
 
 static const uint64_t blake2b_IV[8] =
diff --git a/src/libsodium/crypto_sign/ed25519/ref10/open.c b/src/libsodium/crypto_sign/ed25519/ref10/open.c
index 2731e813..0a7433ad 100644
--- a/src/libsodium/crypto_sign/ed25519/ref10/open.c
+++ b/src/libsodium/crypto_sign/ed25519/ref10/open.c
@@ -3,7 +3,7 @@
 #include <stdint.h>
 #include <string.h>
 
-#include "crypto_hash_sha512.h"
+#include "crypto_generichash.h"
 #include "crypto_sign_ed25519.h"
 #include "crypto_verify_32.h"
 #include "utils.h"
@@ -83,7 +83,7 @@ crypto_sign_ed25519_verify_detached(const unsigned char *sig,
                                     unsigned long long mlen,
                                     const unsigned char *pk)
 {
-    crypto_hash_sha512_state hs;
+    crypto_generichash_state hs;
     unsigned char h[64];
     unsigned char rcheck[32];
     unsigned int  i;
@@ -110,11 +110,11 @@ crypto_sign_ed25519_verify_detached(const unsigned char *sig,
     if (d == 0) {
         return -1;
     }
-    crypto_hash_sha512_init(&hs);
-    crypto_hash_sha512_update(&hs, sig, 32);
-    crypto_hash_sha512_update(&hs, pk, 32);
-    crypto_hash_sha512_update(&hs, m, mlen);
-    crypto_hash_sha512_final(&hs, h);
+    crypto_generichash_init(&hs, NULL, 0, 64);
+    crypto_generichash_update(&hs, sig, 32);
+    crypto_generichash_update(&hs, pk, 32);
+    crypto_generichash_update(&hs, m, mlen);
+    crypto_generichash_final(&hs, h, 64);
     sc_reduce(h);
 
     ge_double_scalarmult_vartime(&R, h, &A, sig + 32);
diff --git a/src/libsodium/crypto_stream/chacha20/ref/stream_chacha20_ref.c b/src/libsodium/crypto_stream/chacha20/ref/stream_chacha20_ref.c
index 15cbae7..42635b1 100644
--- a/src/libsodium/crypto_stream/chacha20/ref/stream_chacha20_ref.c
+++ b/src/libsodium/crypto_stream/chacha20/ref/stream_chacha20_ref.c
@@ -15,6 +15,9 @@
 #include "../stream_chacha20.h"
 #include "../../sodium/common.h"
 
+extern void bsod(const char*,int);
+#define abort() bsod(__FILE__,__LINE__)
+
 struct chacha_ctx {
     uint32_t input[16];
 };
diff --git a/src/libsodium/sodium/utils.c b/src/libsodium/sodium/utils.c
index 5228e40..9d203bf 100644
--- a/src/libsodium/sodium/utils.c
+++ b/src/libsodium/sodium/utils.c
@@ -62,6 +62,9 @@ _sodium_dummy_symbol_to_prevent_memzero_lto(void * const pnt, const size_t len)
 }
 #endif
 
+extern void bsod(const char*,int);
+#define abort() bsod(__FILE__,__LINE__)
+
 void
 sodium_memzero(void * const pnt, const size_t len)
 {
